### PR TITLE
Group pseudo nodes by the node the referred to and their address in grouped topologies.

### LINKED ISSUE #105: All pseudo nodes are grouped together in the "grouped" view
This seems undesirable, implying relationships between unrelated services.

Whats more, they don't have a label in the UI.

Options to fix this include:
- just group by IP
- apply same grouping logic as in #99
- don't show them

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #105

### CODE REVIEW

**Comment by peterbourgon** [comment]
I don't think it's OK to have this behavior be default. Again, I _want_ to see pseudonodes in the general case. The case where Chrome is opening a bunch of short-lived connections is an outlier.

File: report/topology_test.go [hunk 1]
```diff
@@ -231,12 +221,15 @@ func TestRenderByProcessPIDGrouped(t *testing.T) {
...
-			Adjacency:   NewIDList("curl", "localUnknown"),
+			ID:         "apache",
+			LabelMajor: "apache",
+			LabelMinor: "",
+			Rank:       "215",
+			Pseudo:     false,
+			Adjacency: NewIDList(
+				"curl",
+				"pseudo:;10.10.10.10;apache",
+				"pseudo:;10.10.10.11;apache"),
```
Comment (peterbourgon) [inline report/topology_test.go]: We can leverage Go's tolerance of trailing commas to put the closing parens on a newline, making this much nicer.

```go
Adjacency: NewIDList(
	"curl",
	"pseudo:;10.10.10.10;apache",
	"pseudo:;10.10.10.11;apache",
),
```

**Comment by tomwilkie** [comment]
This change is for grouped topologies, and stops all pseudo nodes being merged into one.  So it adds more pseudo nodes....

**Comment by tomwilkie** [comment]
And yes, in general I'm factoring out the pseudo node handling to be per-topology, so different topologies can have different concepts for pseudo nodes.  So for the pid topology, we can show more, for containers, less, and hosts is pretty obvious.

**Comment by peterbourgon** [comment]
Great! I noticed the formatLabel thing is basically dead code at this point, I'll remove that in another PR. :+1:

**Comment by tomwilkie** [comment]
PTAL; I've factored out the pseudo node handling (hopefully with no semantic change)

File: report/mapping_functions.go [hunk 2]
```diff
@@ -25,6 +26,17 @@ type MappedNode struct {
...
+// PseudoFunc is anything which can take an arbitrary node id and produce
+// a pseudo node for the given id, in the form of a node ID and a
+// human-readable major and minor labels.
+//
+// First input param is the src id (before mapping), second is the src rendered node,
+// third is the dst id, fourth is grouped.
+//
+// If the final output parameter is false, the node shall be omitted from the
+// rendered topology.
+type PseudoFunc func(string, RenderableNode, string, bool) (MappedNode, bool)
```
Comment (peterbourgon) [inline report/mapping_functions.go]: How about this?

```go
// PseudoFunc maps arbitrary concrete nodes to renderable pseudonodes.
type PseudoFunc func(srcNodeID string, rn RenderableNode, dstNodeID string, grouped bool) (mn MappedNode, show bool)
```

File: report/topology.go [hunk 3]
```diff
@@ -70,7 +69,7 @@ func NewTopology() Topology {
 //
 // RenderBy takes a a MapFunc, which defines how to group and label nodes. If
 // grouped is true, nodes that belong to the same "class" will be merged.
-func (t Topology) RenderBy(f MapFunc, grouped bool) map[string]RenderableNode {
+func (t Topology) RenderBy(f MapFunc, pseudo PseudoFunc, grouped bool) map[string]RenderableNode {
```
Comment (peterbourgon) [inline report/topology.go]: For consistency...

```go
func (t Topology) RenderBy(map MapFunc, pseudo PseudoFunc, grouped bool) map[string]RenderableNode {
```

**Comment by peterbourgon** [comment]
Yes, this is a brilliant change. :100:

File: report/topology.go (hunk 3 above)
Comment (tomwilkie) [inline report/topology.go]: map is not allowed as a variable name

File: report/mapping_functions.go (hunk 2 above)
Comment (tomwilkie) [inline report/mapping_functions.go]: It doesn't strictly map the renderable nodes to a pseudo node, but a node id to pseudo node (as there is no renderable node by definition).  I've made the role of srcNode more obvious.

**Comment by tomwilkie** [comment]
Okay one more look pls.

**Comment by tomwilkie** [comment]
Also fixes #113

File: report/mapping_functions.go [hunk 4]
```diff
@@ -25,6 +25,12 @@ type MappedNode struct {
 // rendered topology.
 type MapFunc func(string, NodeMetadata, bool) (MappedNode, bool)
 
+// PseudoFunc creates MappedNode representing pseudo nodes given the dstNodeID.
+// The srcNode renderable nodes is essentially from MapFunc, representing one of
```
Comment (peterbourgon) [inline report/mapping_functions.go]: "The srcNode renderable **node** is essentially from MapFunc..."

**Comment by peterbourgon** [comment]
:+1:

**Pull request merged**